### PR TITLE
[DOC] Adding PyTorch / Skorch example

### LINKED ISSUE #664: [DOC] Include DL example with PyTorch and Skorch
### Please confirm that you have searched existing issues in the repository.

- [x] Yes

### Issue with current documentation:

Current documentation lacks an example with a Deep Learning model.

### Idea or request for content:

It would be nice to include one with PyTorch and Skorch simply to show that it's possible to interface the two, on a relevant high dimensional example. Since the dependencies are somewhat heavy especially for the CI, we could rule that example out of the main CI pipeline, and run it not as often, for instance when a release is being made ?

### Additional context (e.g. screenshots, GIFs)

_No response_

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### PR DESCRIPTION
**Pull request by GrituX**
Starting anew from #704 . This closes #664

### CODE REVIEW

File: examples/plot_advanced_skorch.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+"""
+Using PyTorch with HiDimStat
+============================
+HiDimStat was designed with the goal of being easily compatible with Sklearn.
+PyTorch models might not be used directly with HiDimStat for that reason.
+However, with the help of a third party library `Skorch <https://skorch.readthedocs.io/en/stable/>`,
```
Comment (bthirion) [inline examples/plot_advanced_skorch.py]: the link does not render well.

File: examples/plot_advanced_skorch.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+target_fwer = 0.1
+
+X_cluster = resample(
+    X_4_7, n_samples=n_samples // 2, replace=False, random_state=0
+)
+connectivity = image.grid_to_graph(n_x=shape[0], n_y=shape[1])
+clustering = FeatureAgglomeration(
+    n_clusters=n_clusters, connectivity=connectivity
+)
+clustering.fit(X_cluster)
```
Comment (bthirion) [inline examples/plot_advanced_skorch.py:149]: Maybe it would be good to show the "superpixels" returned by the clustering.

File: examples/plot_advanced_skorch.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+_, ax = plt.subplots(
+    1, 2, figsize=(5, 2), subplot_kw={"xticks": [], "yticks": []}
+)
+
+pixel_selection = clustering.inverse_transform(selected_4_7)
+mask_cfi = pixel_selection.reshape(shape)
+importance_cfi_img = clustering.inverse_transform(importances_4_7).reshape(
+    shape
+)
+
```
Comment (bthirion) [inline examples/plot_advanced_skorch.py:209]: Why does the importance image looks so binary ? This is weird.

**Review by bthirion — COMMENTED** [review]
Looks good. Nitpicking on some details.

File: examples/plot_advanced_skorch.py (hunk 2 above)
Comment (GrituX) [reply examples/plot_advanced_skorch.py:149]: I added it there, and replotted it at the end for a better visualization of cluster importance and selection.

File: examples/plot_advanced_skorch.py (hunk 3 above)
Comment (GrituX) [reply examples/plot_advanced_skorch.py:209]: It seems that method `predict_proba` returns values very close to 0 and 1 for each class, there is practically no in-between value. I should check whether Skorch automatically applies a sigmoid during prediction.

Comment (bthirion) [reply examples/plot_advanced_skorch.py:209]: That's OK.

File: examples/plot_advanced_skorch.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+clustering = FeatureAgglomeration(
+    n_clusters=n_clusters, connectivity=connectivity
+)
+clustering.fit(X_cluster)
+
+_, ax = plt.subplots(1, 1, figsize=(2, 2))
+ax.imshow(clustering.labels_.reshape(28, 28), cmap="Set2")
+ax.axis("off")
+ax.set_title("Clustering labels")
+_ = plt.tight_layout()
```
Comment (bthirion) [inline examples/plot_advanced_skorch.py]: since the plot is repeated at the end, we can skip it here ?

**Review by bthirion — COMMENTED** [review]
LGTM otherwise.

**Comment by bthirion** [comment]
6minutes to run it... This will skrew up our CI. I would advocate finding a way to run these examples not systematically. e.g. when explicitly launching a "full build", like before a release.

File: examples/plot_advanced_skorch.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+y_target = y_4_7.astype(np.int64).copy()
+y_target[y_target == 4] = 0
+y_target[y_target == 7] = 1
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X_4_7, y_target, test_size=0.3, random_state=0, stratify=y_target
+)
+model.fit(X_train, y_train)
+print(f"Model accuracy on the test set: {model.score(X_test, y_test):.2f}")
+importances_4_7 = cfi.fit_importance(X_test, y_test)
```
Comment (jpaillard) [inline examples/plot_advanced_skorch.py]: CFI should be fit on the train set
```suggestion
cfi.fit(X_train, y_train)
importances_4_7 = cfi.importance(X_test, y_test)
```

**Review by jpaillard — COMMENTED** [review]
It seems that the forward passes in CFI take quite long. 

I tried a few things to speed up the example: 
 - Reduce `n_permutations` from 20 to 5
 - Parallelize, this can be done by setting `from joblib import parallel_config` and `with parallel_config(backend="threading"):...` and avoid the pickling issues. 
 - cast the data to`.astype(np.float32)`
 - If using CFI: switch `imputation_model_continuous` to `LinearRegression`
 - Switch from CFI to PFI (I would to that last, since it is mostly the forward passes that are slow)
 - remoev the `StandardScaler` from the pipeline (probably marginal gains)

**Comment by GrituX** [comment]
>     * Parallelize, this can be done by setting `from joblib import parallel_config` and `with parallel_config(backend="threading"):...` and avoid the pickling issues.

Thx for this suggestion, this really helped (the n_permutation reduction too). I also reduced the total amount of samples to 500 which should help

**Comment by GrituX** [comment]
The example takes only 36s now which is the 5th longest example to build for the moment. I think this is good now.

**Review by jpaillard — APPROVED** [review]
Looks good. 
Something funny is that the most important cluster is actually the one containing the "background" pixels

File: examples/plot_advanced_skorch.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+mask_cfi = pixel_selection.reshape(shape)
+importance_cfi_img = clustering.inverse_transform(importances_4_7).reshape(
+    shape
+)
+
+cmap = ListedColormap(["tab:red", "white", "tab:blue"])
+ax[0].imshow(mask_cfi, cmap=cmap, vmin=-1, vmax=1)
+ax[0].set_title("4 vs 7 selection", fontweight="bold", y=1.0)
+ax[1].imshow(importance_cfi_img)
+ax[1].set_title("4 vs 7 importance", fontweight="bold", y=1.0)
```
Comment (bthirion) [inline examples/plot_advanced_skorch.py:215]: when we visualize importance, it seems essential to provide a colorbar.

**Review by bthirion — COMMENTED** [review]
LGTM besides one detail.

**Review by bthirion — APPROVED** [review]
LGMT, thx.
